### ISSUE TITLE
Unable to use the dropdown list to select sketches

### ISSUE THREAD

**Issue by AvikaN** [josephtaylor/sketch-mapper#10]
I am getting a runtime error when trying to use the controlP5 dropdown list to select different sketches to run on the quad surfaces: java.lang.RuntimeException: java.lang.ClassCastException: controlP5.DropdownList cannot be cast to controlP5.ControlGroup. This is probably an error with the controlP5 library, but I am using the latest 2.2.5 version with version 3.0 of SketchMapper on Processing 2.2.1. Any idea how to fix the error?

**Comment by AvikaN** [comment]
Has anybody else ever ran into this problem?

**Comment by b2renger** [comment]
I got this error a few times with a class of student. They were indeed using processing 2.2.1 and latest controlp5 

Have you tried using control p5 2.2.4  ? which works fine for me  (again with 2.2.1) 

you can get it here : https://github.com/sojamo/controlp5/releases

and install it manually (the old processing way - by copying the files in sketchbook/library)

**Comment by AvikaN** [comment]
Thanks! I've just tried out your suggestion. Unfortunately it didn't work out. I am trying combinations of different versions of controlP5 and SketchMapper and consistently getting errors that something has gone wrong in controlP5.

**Comment by AvikaN** [comment]
ControlP5 always seems prints out which version of the library the sketch is using. I downloaded version 2.2.4 and imported it the old fashion way. When I run the sketch, its prints out that I am using version 2.2.5. Out of curiosity, is it suppose to print out "ControlP5 2.2.4" ? If so, I am probably getting the wrong version and am unsure about how to get access to version 2.2.4.

**Comment by b2renger** [comment]
The detection of the lib depends on the folder hierarchy.

If I have to deal with several version I just rename the folder of the version I don't want to use to something else, for instance "controlP5-off"

**Comment by isaacmuroxtec** [comment]
I solved the problem change one line in SketchMapper.java (line 157): 

```
                //if (e.getGroup().getCaptionLabel().getText().equals(sketch.getName())) {
                if (e.getController().getLabel().equals(sketch.getName())) {
```

and rebuild the library. It works for me. 
I'am using ControlP5 2.2.5, Processing 2.2.1, SketchMapper 2.0.3 and Ubuntu 12.04.

**Comment by josephtaylor** [comment]
thanks @willezumleben for that fix. and sorry for the delay ! released fix in 3.0.2 for Processing 2 and 4.0.0 for Processing 3.

**Comment by ghost** [comment]
This is still an issue in Mac OSX both in Processing 2.2.1 and Processing 3.0.2 I am unable to use the dropdown to select sketches. Whenever I try to assign a new sketch to a surface and click on the dropdown the surface gets unselected. Using ControlP5 2.2.5

**Comment by ghost** [comment]
Just tested it on Windows 7, too. Processing 2.2.1, Control P5 2.2.5, SketchMapper 3.0.2 -> dropdown to assign sketch to surface not working.

**Comment by ghost** [comment]
solved it using ScrollableLists instead of DropdownLists and replacing the switch statemen code for case 9 with the following code:

`
case 9:                                                                                                                                  
    for (Sketch sketch : surfaceMapper.getSketchList()) {                                                                                                                                                                                                                        
        int index =  (int)e.getController().getValue();                                                                                  
        String sketchSelected = controlP5.get(ScrollableList.class,e.getController().getName()).getItem(index).get("name").toString();  
        //PApplet.println(sketchSelected);                                                                                                 
        if (sketchSelected.equals(sketch.getName())) {                                                                                   
            //PApplet.println("setting new sketch to surface : " + surfaceMapper.getSurfaces().get(mostRecentSurface).getId());          
            surfaceMapper.getSurfaces().get(mostRecentSurface).setSketch(sketch);                                                        
            break;                                                                                                                       
        }                                                                                                                                
    }                                                                                                                                    
    break; 
`

hope it helps.

best,

paulo

**Comment by ghost** [comment]
in case you are interested, I added my working version for Processing 3.1 and controlP5 2.25 here https://github.com/pauloguerraf/sketch-mapper
using IntelliJ as development environment...

**Comment by josephtaylor** [comment]
@pauloguerraf I appreciate all your help, gonna look into incorporating your fix as soon as I can.

**Comment by josephtaylor** [comment]
So, ControlP5 came out with a new version, and I updated to that.
Did a couple tests with controlP5 2.2.6, Processing 3.1 on Linux and everything seems to work fine.
I don't have a windows or mac machine right now to test it out on.

Just released 4.1.1 that incorporates the new ControlP5 version, can you give that a shot ?
